plugin/pkg/shoot/validator/admission.go

```diff
@@ -2040,3 +2043,65 @@ func isShootInMigrationOrRestorePhase(shoot *core.Shoot) bool {
 			shoot.Status.LastOperation.State != core.LastOperationStateSucceeded ||
 			shoot.Status.LastOperation.Type == core.LastOperationTypeMigrate)
 }
+
+func (c *validationContext) validateManagedServiceAccountIssuer(
+	a admission.Attributes,
+	secretLister kubecorev1listers.SecretLister,
+) error {
+	// Skip the validation if no managed service account issuer configuration is involved.
+	if !helper.HasManagedIssuer(c.shoot) &&
+		(c.oldShoot == nil || !helper.HasManagedIssuer(c.oldShoot)) {
+		return nil
+	}
+
+	managedIssuerConfigSecrets, err := secretLister.Secrets(v1beta1constants.GardenNamespace).List(labels.SelectorFromSet(labels.Set{
+		v1beta1constants.GardenRole: v1beta1constants.GardenRoleShootServiceAccountIssuer,
+	}))
+	if err != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not retrieve managed service account issuer config secret: %+v", err.Error()))
+	}
+
+	// Skip the validation if no managed service account issuer secrets are found and shoot does not contain managed service account issuer configs.
+	if len(managedIssuerConfigSecrets) == 0 {
+		// Shoots should not be allowed to enable feature that is not configured for the Gardener installation.
+		if helper.HasManagedIssuer(c.shoot) {
+			return admission.NewForbidden(a, errors.New("cannot enable managed service account issuer as it is not supported in this Gardener installation"))
+		}
+		// Old shoot object has the feature enabled but for some reason the configuration is missing.
+		// Reconciliation of such cluster can break existing integrations.
+		// An intervention from a Gardener admin/operator is required.
+		if c.oldShoot != nil && helper.HasManagedIssuer(c.oldShoot) {
+			return apierrors.NewInternalError(errors.New("old shoot object has managed service account issuer enabled, but Gardener configuration is missing"))
+		}
+		return nil
+	}
+
+	// The managed service account issuer is configured centrally per Garden cluster.
+	// The presence of more than one secret is an ambiguous behaviour and should be disallowed.
+	if len(managedIssuerConfigSecrets) > 1 {
+		return apierrors.NewInternalError(fmt.Errorf("can only accept at most one managed service account issuer secret, but found %d", len(managedIssuerConfigSecrets)))
+	}
+
+	// Fail the validation if secret is present, but not configured correctly.
+	configSecret := managedIssuerConfigSecrets[0]
```

Comment:
The number of secrets as well as the content is already checked in [ReadGardenSecrets](https://github.com/gardener/gardener/blob/a5b8a76bdd1a69fa54759996bd9662ac47ab954d/pkg/utils/gardener/garden.go#L101), and I think it's not really required to maintain this duplicated logic here. DNS secrets are treated the same way.
If you think there is are very good reasons to have these checks here as well, I'd suggest to make the validation logic reusable.

Reply:
I removed the duplication